File: AGTIA_Objective-C/AGTIACircleView/AGTIACircleView/CircleView/CircleLayer.m
//
//  CircleLayer.m
//  AGTIACircleView
//
//  Created by fernando on 2017/2/3.
//  Copyright © 2017年 fernando. All rights reserved.
//

#import "CircleLayer.h"
#import <UIKit/UIKit.h>

typedef NS_ENUM(NSInteger, MovingPoint){
    Point_B,
    Point_D
};

// 外接矩形的边长
#define outsideRectSize 90

@interface CircleLayer ()
// 外接矩形的尺寸
@property (nonatomic, assign) CGRect outSideRect;
// 上次的progress， 方便做差得出拖动方向
@property (nonatomic, assign) CGFloat lastProgress;
// 实时记录滑动方向
@property (nonatomic, assign) MovingPoint movingPoint;

@end

@implementation CircleLayer

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.lastProgress = 0.5;
    }
    return self;
}

- (instancetype)initWithLayer:(CircleLayer *)layer
{
    self = [super initWithLayer:layer];
    if (self) {
        self.outSideRect = layer.outSideRect;
        self.lastProgress = layer.lastProgress;
        self.movingPoint = layer.movingPoint;
    }
    return self;
}

- (void)setProgress:(CGFloat)progress {
    _progress = progress;
    
    // 判断拖动的方向
    if (progress <= 0.5) {  // 外接矩形在左，则改变B点
        self.movingPoint = Point_B;
        NSLog(@"B点动");
    } else { // 外接矩形在右，则改变D点
        self.movingPoint = Point_D;
        NSLog(@"D点动");
    }
    
    self.lastProgress = progress;
    
    // 计算外接矩阵的位置大小
    CGFloat x = self.position.x - outsideRectSize/2 + (progress - 0.5)*(self.frame.size.width - outsideRectSize);
    CGFloat y = self.position.y - outsideRectSize/2;
    
    self.outSideRect = CGRectMake(x, y, outsideRectSize, outsideRectSize);
    
    // 开始绘制
    [self setNeedsDisplay];
}

- (void)drawInContext:(CGContextRef)ctx {
    // “offset 指的是 A-C1,B-C2... 的距离，当该值设置为正方形边长的 1/3.6 倍时，画出来的圆弧近似贴合 1/4 圆。为什么是 3.6 ？这里 有一篇文章。文章里三阶贝塞尔曲线拟合 1/4 圆的时候最佳参数 h=0.552,  表示的意义是：当正方形边长的 1/2 为 1 （ 即正方形边长为 2） 时， offset  等于 0.552  就能使圆弧近似贴近 1/4 圆。所以比例系数为 1/0.552 ，即正方形边长和 offset 的比例系数为：2/0.552 = 3.623。近似于 3.6。其实还有种更直观的近似方法：如果圆心为 O，OC1, OC2  就一定是三等分点，也就是夹角为 30°，那么 AC1 （也就是 offset  ）就等于 1/2的边长 *  tan30° 。
    // “顺便解释两个地方。1.为什么要*2? 因为为了让  D 点区别于 A 点，让 D 移动距离比 A 多，你完全可以 *3 ,*2.5，但是不能移动太远。2.为什么要是1/6？这个 1/6 也是自己定的。你可以让 A 移动 1/7 ，1/10 都可以，但是最好不要太靠近 1/2，这时 A 点就移到中点了，形变的样子就不好看了。”
    // A-C1, B-C2 ... 的距离，当设置为正方形边长的1/3.6时，画出的圆弧完美贴合圆形
    CGFloat offset = self.outSideRect.size.width / 3.6;
    
    // A.B.C.D实际需要移动的距离.系数为滑块偏离中点0.5的绝对值乘以2.当滑到两端的时候，movedDistance为最大值:「外接矩形宽度的1/6」
    CGFloat movedDistance = self.outSideRect.size.width / 6 * fabs(self.progress - 0.5)*2;
    
    // 方便下方计算各点坐标，先计算出外接矩形的中心点坐标
    CGPoint rectCenter = CGPointMake(self.outSideRect.origin.x + self.outSideRect.size.width/2, self.outSideRect.origin.y + self.outSideRect.size.height/2);
    
    CGPoint pointA = CGPointMake(rectCenter.x, self.outSideRect.origin.y + movedDistance);
    CGPoint pointB = CGPointMake(self.movingPoint == Point_D ? rectCenter.x + self.outSideRect.size.width/2 : rectCenter.x + self.outSideRect.size.width/2 + movedDistance * 2, rectCenter.y);
    CGPoint pointC = CGPointMake(rectCenter.x, rectCenter.y + self.outSideRect.size.height/2 - movedDistance);
    CGPoint pointD = CGPointMake(self.movingPoint == Point_D ? rectCenter.x - self.outSideRect.size.width/2 - movedDistance * 2 : self.outSideRect.origin.x, rectCenter.y);

    
    CGPoint c1 = CGPointMake(pointA.x + offset, pointA.y);
    CGPoint c2 = CGPointMake(pointB.x, self.movingPoint == Point_D ? pointB.y - offset : pointB.y - offset + movedDistance);
    CGPoint c3 = CGPointMake(pointB.x, self.movingPoint == Point_D ? pointB.y + offset : pointB.y + offset - movedDistance);
    CGPoint c4 = CGPointMake(pointC.x + offset, pointC.y);
    CGPoint c5 = CGPointMake(pointC.x - offset, pointC.y);
    CGPoint c6 = CGPointMake(pointD.x, self.movingPoint == Point_D ? pointD.y + offset - movedDistance: pointD.y + offset);
    CGPoint c7 = CGPointMake(pointD.x, self.movingPoint == Point_D ? pointD.y - offset + movedDistance: pointD.y - offset);
    CGPoint c8 = CGPointMake(pointA.x - offset, pointA.y);
    
    // 外接虚线矩形
    UIBezierPath *rectPath = [UIBezierPath bezierPathWithRect:self.outSideRect];
    CGContextAddPath(ctx, rectPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor blackColor].CGColor);
    CGContextSetLineWidth(ctx, 1.0);
    CGFloat dash[] = {5.0, 5.0};
    CGContextSetLineDash(ctx, 0.0, dash, 2);
    CGContextStrokePath(ctx);
    
    
    // 圆的边界
    UIBezierPath *ovalPath = [UIBezierPath bezierPath];
    [ovalPath moveToPoint:pointA];
    [ovalPath addCurveToPoint:pointB controlPoint1:c1 controlPoint2:c2];
    [ovalPath addCurveToPoint:pointC controlPoint1:c3 controlPoint2:c4];
    [ovalPath addCurveToPoint:pointD controlPoint1:c5 controlPoint2:c6];
    [ovalPath addCurveToPoint:pointA controlPoint1:c7 controlPoint2:c8];
    [ovalPath closePath];
    
    CGContextAddPath(ctx, ovalPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor blackColor].CGColor);
    CGContextSetFillColorWithColor(ctx, [UIColor redColor].CGColor);
    CGContextSetLineDash(ctx, 0.0, NULL, 0.0);
    CGContextDrawPath(ctx, kCGPathFillStroke);
    
    // 标记出每个点并且连接， 方便观察， 给所有关键点染黄色， 所有辅助线，黑色
    CGContextSetFillColorWithColor(ctx, [UIColor yellowColor].CGColor);
    CGContextSetStrokeColorWithColor(ctx, [UIColor blackColor].CGColor);
    NSArray *values = @[[NSValue valueWithCGPoint:pointA],[NSValue valueWithCGPoint:pointB],[NSValue valueWithCGPoint:pointC],[NSValue valueWithCGPoint:pointD],[NSValue valueWithCGPoint:c1],[NSValue valueWithCGPoint:c2],[NSValue valueWithCGPoint:c3],[NSValue valueWithCGPoint:c4],[NSValue valueWithCGPoint:c5],[NSValue valueWithCGPoint:c6],[NSValue valueWithCGPoint:c7],[NSValue valueWithCGPoint:c8]];
    [self drawPoint:values withContext:ctx];
    
    // 连接辅助线
    UIBezierPath *helperPath = [UIBezierPath bezierPath];
    [helperPath moveToPoint:pointA];
    [helperPath addLineToPoint:c1];
    [helperPath addLineToPoint:c2];
    [helperPath addLineToPoint:pointB];
    [helperPath addLineToPoint:c3];
    [helperPath addLineToPoint:c4];
    [helperPath addLineToPoint:pointC];
    [helperPath addLineToPoint:c5];
    [helperPath addLineToPoint:c6];
    [helperPath addLineToPoint:pointD];
    [helperPath addLineToPoint:c7];
    [helperPath addLineToPoint:c8];
    [helperPath closePath];
    
    CGContextAddPath(ctx, helperPath.CGPath);
    CGFloat dash2[] = {2.0, 2.0};
    CGContextSetLineDash(ctx, 0, dash2, 2);
    CGContextStrokePath(ctx);
}


//ctx字面意思是上下文，你可以理解为一块全局的画布。也就是说，一旦在某个地方改了画布的一些属性，其他任何使用画布的属性的时候都是改了之后的。比如上面在 //1 中把线条样式改成了虚线，那么在下文 //2 中如果不恢复成连续的直线，那么画出来的依然是//1中的虚线样式。

// 在某个point位置画一个点，方便观察运动情况
- (void)drawPoint:(NSArray *)points withContext:(CGContextRef)ctx {
    for (NSValue *pointValue in points) {
        CGPoint point = pointValue.CGPointValue;
        CGContextFillRect(ctx, CGRectMake(point.x - 2, point.y - 2, 4, 4));
    }
}


@end

File: AGTIA_Swift/AnimatedCircle/AnimatedCircle/CircleView/CircleLayer.swift
//
//  CircleLayer.swift
//  AnimatedCircle
//
//  Created by fernando on 2017/2/3.
//  Copyright © 2017年 fernando. All rights reserved.
//

import UIKit

enum MovingPoint {
    case Point_B
    case Point_D
}

// 外接矩形的长(宽)
let outsideRectSize: CGFloat = 90

class CircleLayer: CALayer {
    var progress: CGFloat = 0.0 {
        didSet {
            // 外接矩形在左侧，则改变B点，反之在右边，则改变D点
            if progress <= 0.5 {
                movePoint = .Point_B
                print("B点动")
            } else {
                movePoint = .Point_D
                print("D点动")
            }
            
            self.lastProgress = progress
            
            let buff = (progress - 0.5) * (frame.size.width - outsideRectSize)
            let x = position.x - outsideRectSize/2 + buff
            let y = position.y - outsideRectSize/2
            outsideRect = CGRect(x: x, y: y, width: outsideRectSize, height: outsideRectSize)
            
            setNeedsDisplay()
        }
    }
    
    private var outsideRect: CGRect!
    private var lastProgress: CGFloat = 0.5
    private var movePoint: MovingPoint!
    
    override init() {
        super.init()
    }
    
    override init(layer: Any) {
        super.init(layer: layer)
        if let layer = layer as? CircleLayer {
            progress = layer.progress
            lastProgress = layer.lastProgress
            outsideRect = layer.outsideRect
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    override func draw(in ctx: CGContext) {

        let offset = outsideRect.size.width / 3.6
        let movedDistance = (outsideRect.size.width / 6) * fabs(self.progress - 0.5) * 2
        let rectCenter = CGPoint(x: outsideRect.origin.x + outsideRect.size.width/2, y: outsideRect.origin.y + outsideRect.size.height/2)
        
        
        let pointA = CGPoint(x: rectCenter.x, y: outsideRect.origin.y + movedDistance)
        let pointB = CGPoint(x: movePoint == .Point_D ? rectCenter.x + outsideRect.size.width/2 : rectCenter.x + outsideRect.size.width/2 + movedDistance * 2, y: rectCenter.y)
        let pointC = CGPoint(x: rectCenter.x, y: rectCenter.y + outsideRect.size.height/2 - movedDistance)
        let pointD = CGPoint(x: movePoint == .Point_D ? rectCenter.x - outsideRect.size.width/2 - movedDistance * 2 : rectCenter.x - outsideRect.size.width/2, y: rectCenter.y)
    
        
        let c1 = CGPoint(x: pointA.x + offset, y: pointA.y)
        let c2 = CGPoint(x: pointB.x, y: movePoint == .Point_D ? pointB.y - offset : pointB.y - offset + movedDistance)
        let c3 = CGPoint(x: pointB.x, y: movePoint == .Point_D ? pointB.y + offset : pointB.y + offset - movedDistance)
        let c4 = CGPoint(x: pointC.x + offset, y: pointC.y)
        let c5 = CGPoint(x: pointC.x - offset, y: pointC.y)
        let c6 = CGPoint(x: pointD.x, y: movePoint == .Point_D ? pointD.y + offset - movedDistance : pointD.y + offset)
        let c7 = CGPoint(x: pointD.x, y: movePoint == .Point_D ? pointD.y - offset + movedDistance : pointD.y - offset)
        let c8 = CGPoint(x: pointA.x - offset, y: pointA.y)
        
        // 绘制外接虚线矩形
        let rectPath = UIBezierPath(rect: outsideRect)
        ctx.addPath(rectPath.cgPath)
        ctx.setStrokeColor(UIColor.black.cgColor)
        ctx.setLineWidth(1.0)
        let dash = [CGFloat(5.0), CGFloat(5.0)]
        ctx.setLineDash(phase: 0, lengths: dash)
        ctx.strokePath()

        
        // 圆的边界
        let ovalPath = UIBezierPath()
        ovalPath.move(to: pointA)
        ovalPath.addCurve(to: pointB, controlPoint1: c1, controlPoint2: c2)
        ovalPath.addCurve(to: pointC, controlPoint1: c3, controlPoint2: c4)
        ovalPath.addCurve(to: pointD, controlPoint1: c5, controlPoint2: c6)
        ovalPath.addCurve(to: pointA, controlPoint1: c7, controlPoint2: c8)
        ovalPath.close()
        
        ctx.addPath(ovalPath.cgPath)
        ctx.setStrokeColor(UIColor.black.cgColor)
        ctx.setFillColor(UIColor.blue.cgColor)
        ctx.setLineDash(phase: 0, lengths: [])
        ctx.drawPath(using: .fillStroke)
        
        //标记出每个点并连线，方便观察，给所有关键点染色 -- 黄色,辅助线颜色 -- 绿色
        ctx.setFillColor(UIColor.yellow.cgColor)
        ctx.setStrokeColor(UIColor.green.cgColor)
        let points = [NSValue(cgPoint: pointA),NSValue(cgPoint: c1),NSValue(cgPoint: c2),NSValue(cgPoint: pointB),NSValue(cgPoint: c3),NSValue(cgPoint: c4),NSValue(cgPoint: pointC),NSValue(cgPoint: c5),NSValue(cgPoint: c6),NSValue(cgPoint: pointD),NSValue(cgPoint: c7),NSValue(cgPoint: c8)];
        // 填充点
        drawPoint(points: points, ctx: ctx)
        // 连接点
        let helperLine = UIBezierPath()
        helperLine.move(to: pointA)
        helperLine.addLine(to: c1)
        helperLine.addLine(to: c2)
        helperLine.addLine(to: pointB)
        helperLine.addLine(to: c3)
        helperLine.addLine(to: c4)
        helperLine.addLine(to: pointC)
        helperLine.addLine(to: c5)
        helperLine.addLine(to: c6)
        helperLine.addLine(to: pointD)
        helperLine.addLine(to: c7)
        helperLine.addLine(to: c8)
        helperLine.addLine(to: pointA)
        
        ctx.addPath(helperLine.cgPath)
        let dash2 = [CGFloat(2.0), CGFloat(2.0)]
        ctx.setLineDash(phase: 0, lengths: dash2)
        ctx.strokePath()
    }
    
    private func drawPoint(points: [NSValue], ctx: CGContext) {
        for pointValue in points {
            let point = pointValue.cgPointValue
            ctx.fill(CGRect(x: point.x - 2, y: point.y - 2, width: 4, height: 4))
        }
    }
}
